### PR TITLE
i#2820 trace thread subset: reset callbacks at exit

### PR DESCRIPTION
**Pull request by derekbruening**
Resets callback functions to empty at exit to better support re-attach.
Unfortunately i#2175 still prevents extending the thread filter test to
include re-attach.

Issue: #2820

### CODE REVIEW

File: clients/drcachesim/tracer/tracer.cpp [hunk 1]
```diff
@@ -1559,6 +1559,12 @@ event_exit(void)
     if (file_ops_func.exit_cb != NULL)
         (*file_ops_func.exit_cb)(file_ops_func.exit_arg);
 
+    /* Clear callbacks to support reset. */
+    memset(&file_ops_func, 0, sizeof(file_ops_func));
+    if (!offline_instru_t::custom_module_data(nullptr, nullptr, nullptr))
+        DR_ASSERT(false && "failed to clear custom module callbacks");
+    should_trace_thread_cb = nullptr;
```
Comment (Carrotman42) [inline clients/drcachesim/tracer/tracer.cpp]: What about `thread_filtering_enabled`, and `trace_thread_cb_user_data` just in case?

Comment (derekbruening) [reply clients/drcachesim/tracer/tracer.cpp]: Agreed, this is a heavyweight client and a few stores won't matter.  Done.

**Comment by derekbruening** [comment]
@Carrotman42, PTAL: tests should go green (took a bit of work; had to fix #2845 too).

**Comment by Carrotman42** [comment]
LGTM

**Pull request merged**